File: apps/ios/Sources/Importing.swift
import Foundation

enum PracticeKind: String {
    case verb
    case adjective
}

struct ImportFailure: Error, Equatable {
    let message: String
}

enum Importing {
    static func normalizeImport(_ items: [ImportItem], practice: PracticeKind) -> Result<[CardFixture], ImportFailure> {
        var bank: [CardFixture] = []

        for item in items {
            switch item {
            case .string(let dictRaw):
                let dict = dictRaw.trimmingCharacters(in: .whitespacesAndNewlines)
                if dict.isEmpty { return .failure(ImportFailure(message: "存在空的項目。")) }
                switch practice {
                case .verb:
                    let group = Conjugation.inferVerbGroup(dict)
                    do {
                        let conjugated = try Conjugation.conjugateVerb(dict: dict, group: group)
                        bank.append(CardFixture(dict: dict, nai: conjugated.nai, ta: conjugated.ta, nakatta: conjugated.nakatta, te: conjugated.te, potential: conjugated.potential, group: group.rawValue, zh: nil))
                    } catch {
                        return .failure(ImportFailure(message: "無法推導：\(dict)"))
                    }
                case .adjective:
                    let group = Conjugation.inferAdjectiveGroup(dict)
                    do {
                        let conjugated = try Conjugation.conjugateAdjective(dict: dict, group: group)
                        let normalizedDict = conjugated.dict ?? Conjugation.normalizeAdjectiveDict(dict)
                        bank.append(CardFixture(dict: normalizedDict, nai: conjugated.nai, ta: conjugated.ta, nakatta: conjugated.nakatta, te: conjugated.te, potential: nil, group: group.rawValue, zh: nil))
                    } catch {
                        return .failure(ImportFailure(message: "無法推導：\(dict)"))
                    }
                }

            case .object(let record):
                let dict = record.dict.trimmingCharacters(in: .whitespacesAndNewlines)
                if dict.isEmpty { return .failure(ImportFailure(message: "每筆資料需包含 dict。")) }

                switch practice {
                case .verb:
                    if let card = verbCardFromRecord(record) {
                        bank.append(card)
                        continue
                    }
                    let group = VerbGroup(rawValue: record.group ?? "") ?? Conjugation.inferVerbGroup(dict)
                    do {
                        let conjugated = try Conjugation.conjugateVerb(dict: dict, group: group)
                        let overrides = buildOverrides(record, fields: ["nai", "ta", "nakatta", "te", "potential", "zh"])
                        let card = CardFixture(
                            dict: dict,
                            nai: overrides.nai ?? conjugated.nai,
                            ta: overrides.ta ?? conjugated.ta,
                            nakatta: overrides.nakatta ?? conjugated.nakatta,
                            te: overrides.te ?? conjugated.te,
                            potential: overrides.potential ?? conjugated.potential,
                            group: group.rawValue,
                            zh: overrides.zh
                        )
                        bank.append(card)
                    } catch {
                        return .failure(ImportFailure(message: "無法推導：\(dict)"))
                    }

                case .adjective:
                    if let card = adjectiveCardFromRecord(record) {
                        bank.append(card)
                        continue
                    }
                    let group = AdjectiveGroup(rawValue: record.group ?? "") ?? Conjugation.inferAdjectiveGroup(dict)
                    do {
                        let conjugated = try Conjugation.conjugateAdjective(dict: dict, group: group)
                        let overrides = buildOverrides(record, fields: ["nai", "ta", "nakatta", "te", "zh"])
                        let normalizedDict = conjugated.dict ?? Conjugation.normalizeAdjectiveDict(dict)
                        let card = CardFixture(
                            dict: normalizedDict,
                            nai: overrides.nai ?? conjugated.nai,
                            ta: overrides.ta ?? conjugated.ta,
                            nakatta: overrides.nakatta ?? conjugated.nakatta,
                            te: overrides.te ?? conjugated.te,
                            potential: nil,
                            group: group.rawValue,
                            zh: overrides.zh
                        )
                        bank.append(card)
                    } catch {
                        return .failure(ImportFailure(message: "無法推導：\(dict)"))
                    }
                }
            }
        }

        return .success(bank)
    }

    private static func verbCardFromRecord(_ record: ImportObject) -> CardFixture? {
        guard let group = record.group,
              let nai = record.nai,
              let ta = record.ta,
              let nakatta = record.nakatta,
              let te = record.te,
              let potential = record.potential
        else { return nil }
        return CardFixture(dict: record.dict, nai: nai, ta: ta, nakatta: nakatta, te: te, potential: potential, group: group, zh: record.zh)
    }

    private static func adjectiveCardFromRecord(_ record: ImportObject) -> CardFixture? {
        guard let group = record.group,
              let nai = record.nai,
              let ta = record.ta,
              let nakatta = record.nakatta,
              let te = record.te
        else { return nil }
        return CardFixture(dict: record.dict, nai: nai, ta: ta, nakatta: nakatta, te: te, potential: nil, group: group, zh: record.zh)
    }

    private static func buildOverrides(_ record: ImportObject, fields: [String]) -> OverrideValues {
        var overrides = OverrideValues()
        fields.forEach { field in
            let value: String?
            switch field {
            case "nai": value = record.nai
            case "ta": value = record.ta
            case "nakatta": value = record.nakatta
            case "te": value = record.te
            case "potential": value = record.potential
            case "zh": value = record.zh
            default: value = nil
            }
            guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else { return }
            switch field {
            case "nai": overrides.nai = trimmed
            case "ta": overrides.ta = trimmed
            case "nakatta": overrides.nakatta = trimmed
            case "te": overrides.te = trimmed
            case "potential": overrides.potential = trimmed
            case "zh": overrides.zh = trimmed
            default: break
            }
        }
        return overrides
    }

    private struct OverrideValues {
        var nai: String?
        var ta: String?
        var nakatta: String?
        var te: String?
        var potential: String?
        var zh: String?
    }
}

File: apps/android/app/src/main/java/com/learnjapanese/app/data/Importing.kt
package com.learnjapanese.app.data

sealed class ImportResult {
    data class Success(
        val bank: List<CardFixture>,
    ) : ImportResult()

    data class Failure(
        val message: String,
    ) : ImportResult()
}

enum class PracticeKind(
    val raw: String,
) {
    VERB("verb"),
    ADJECTIVE("adjective"),
}

enum class AnswerMode { INPUT, CHOICE }

enum class QuestionType(
    val label: String,
) {
    NAI("ない形"),
    TA("た形"),
    NAKATTA("なかった形"),
    TE("て形"),
    POTENTIAL("可能形"),
    MIXED("混合"),
}

enum class VerbScope(
    val label: String,
) {
    ALL("全部"),
    GODAN("五段"),
    ICHIDAN("二段"),
    IRREGULAR("不規則"),
}

enum class AdjectiveScope(
    val label: String,
) {
    ALL("全部"),
    I("い形"),
    NA("な形"),
}

enum class PracticeMode { NORMAL, REVIEW_WRONG }

object Importing {
    fun normalizeImport(
        items: List<ImportItem>,
        practice: PracticeKind,
    ): ImportResult {
        val bank = mutableListOf<CardFixture>()

        for (item in items) {
            when (item) {
                is ImportItem.StringItem -> {
                    val dict = item.value.trim()
                    if (dict.isEmpty()) return ImportResult.Failure("存在空的項目。")
                    when (practice) {
                        PracticeKind.VERB -> {
                            val group = Conjugation.inferVerbGroup(dict)
                            try {
                                val conjugated = Conjugation.conjugateVerb(dict, group)
                                bank.add(
                                    CardFixture(
                                        dict = dict,
                                        nai = conjugated.nai,
                                        ta = conjugated.ta,
                                        nakatta = conjugated.nakatta,
                                        te = conjugated.te,
                                        potential = conjugated.potential,
                                        group = group.name.lowercase(),
                                        zh = null,
                                    ),
                                )
                            } catch (_: Exception) {
                                return ImportResult.Failure("無法推導：$dict")
                            }
                        }

                        PracticeKind.ADJECTIVE -> {
                            val group = Conjugation.inferAdjectiveGroup(dict)
                            try {
                                val conjugated = Conjugation.conjugateAdjective(dict, group)
                                val normalizedDict = conjugated.dict ?: Conjugation.normalizeAdjectiveDict(dict)
                                bank.add(
                                    CardFixture(
                                        dict = normalizedDict,
                                        nai = conjugated.nai,
                                        ta = conjugated.ta,
                                        nakatta = conjugated.nakatta,
                                        te = conjugated.te,
                                        potential = null,
                                        group = group.name.lowercase(),
                                        zh = null,
                                    ),
                                )
                            } catch (_: Exception) {
                                return ImportResult.Failure("無法推導：$dict")
                            }
                        }
                    }
                }

                is ImportItem.ObjectItem -> {
                    val record = item.value
                    val dict = record.dict.trim()
                    if (dict.isEmpty()) return ImportResult.Failure("每筆資料需包含 dict。")

                    when (practice) {
                        PracticeKind.VERB -> {
                            verbCardFromRecord(record)?.let {
                                bank.add(it)
                                continue
                            }
                            val group =
                                record.group?.let { value ->
                                    runCatching { Conjugation.VerbGroup.valueOf(value.uppercase()) }.getOrNull()
                                } ?: Conjugation.inferVerbGroup(dict)
                            try {
                                val conjugated = Conjugation.conjugateVerb(dict, group)
                                val overrides =
                                    buildOverrides(record, listOf("nai", "ta", "nakatta", "te", "potential", "zh"))
                                bank.add(
                                    CardFixture(
                                        dict = dict,
                                        nai = overrides.nai ?: conjugated.nai,
                                        ta = overrides.ta ?: conjugated.ta,
                                        nakatta = overrides.nakatta ?: conjugated.nakatta,
                                        te = overrides.te ?: conjugated.te,
                                        potential = overrides.potential ?: conjugated.potential,
                                        group = group.name.lowercase(),
                                        zh = overrides.zh,
                                    ),
                                )
                            } catch (_: Exception) {
                                return ImportResult.Failure("無法推導：$dict")
                            }
                        }

                        PracticeKind.ADJECTIVE -> {
                            adjectiveCardFromRecord(record)?.let {
                                bank.add(it)
                                continue
                            }
                            val group =
                                record.group?.let { value ->
                                    runCatching { Conjugation.AdjectiveGroup.valueOf(value.uppercase()) }.getOrNull()
                                } ?: Conjugation.inferAdjectiveGroup(dict)
                            try {
                                val conjugated = Conjugation.conjugateAdjective(dict, group)
                                val overrides = buildOverrides(record, listOf("nai", "ta", "nakatta", "te", "zh"))
                                val normalizedDict = conjugated.dict ?: Conjugation.normalizeAdjectiveDict(dict)
                                bank.add(
                                    CardFixture(
                                        dict = normalizedDict,
                                        nai = overrides.nai ?: conjugated.nai,
                                        ta = overrides.ta ?: conjugated.ta,
                                        nakatta = overrides.nakatta ?: conjugated.nakatta,
                                        te = overrides.te ?: conjugated.te,
                                        potential = null,
                                        group = group.name.lowercase(),
                                        zh = overrides.zh,
                                    ),
                                )
                            } catch (_: Exception) {
                                return ImportResult.Failure("無法推導：$dict")
                            }
                        }
                    }
                }
            }
        }

        return ImportResult.Success(bank)
    }

    private fun verbCardFromRecord(record: ImportObject): CardFixture? {
        val group = record.group ?: return null
        val nai = record.nai ?: return null
        val ta = record.ta ?: return null
        val nakatta = record.nakatta ?: return null
        val te = record.te ?: return null
        val potential = record.potential ?: return null
        return CardFixture(
            dict = record.dict,
            nai = nai,
            ta = ta,
            nakatta = nakatta,
            te = te,
            potential = potential,
            group = group,
            zh = record.zh,
        )
    }

    private fun adjectiveCardFromRecord(record: ImportObject): CardFixture? {
        val group = record.group ?: return null
        val nai = record.nai ?: return null
        val ta = record.ta ?: return null
        val nakatta = record.nakatta ?: return null
        val te = record.te ?: return null
        return CardFixture(
            dict = record.dict,
            nai = nai,
            ta = ta,
            nakatta = nakatta,
            te = te,
            potential = null,
            group = group,
            zh = record.zh,
        )
    }

    private data class OverrideValues(
        var nai: String? = null,
        var ta: String? = null,
        var nakatta: String? = null,
        var te: String? = null,
        var potential: String? = null,
        var zh: String? = null,
    )

    private fun buildOverrides(
        record: ImportObject,
        fields: List<String>,
    ): OverrideValues {
        val overrides = OverrideValues()
        for (field in fields) {
            val value =
                when (field) {
                    "nai" -> record.nai
                    "ta" -> record.ta
                    "nakatta" -> record.nakatta
                    "te" -> record.te
                    "potential" -> record.potential
                    "zh" -> record.zh
                    else -> null
                }
            val trimmed = value?.trim().orEmpty()
            if (trimmed.isEmpty()) continue
            when (field) {
                "nai" -> overrides.nai = trimmed
                "ta" -> overrides.ta = trimmed
                "nakatta" -> overrides.nakatta = trimmed
                "te" -> overrides.te = trimmed
                "potential" -> overrides.potential = trimmed
                "zh" -> overrides.zh = trimmed
            }
        }
        return overrides
    }
}
